File: src/hyper_markdown/imports.py
"""The ``from <ref> import <names>`` mini-grammar (HMD-0001 §5.3).

A one-line statement form rather than nested YAML: the grammar is trivial and
the statement reads at a glance, which nested `from:`/`names:` keys would not.
"""

from __future__ import annotations

import re

from .model import ImportStmt

_STMT_RE = re.compile(r"^from\s+(?P<ref>\S+)\s+import\s+(?P<names>.+?)\s*$")
_NAME_RE = re.compile(r"^(?P<name>[^\s/]+)(?:\s+as\s+(?P<alias>[^\s/]+))?$")


class ImportError_(ValueError):
    """A malformed import statement (HMD014)."""


def is_qualified(ref: str) -> bool:
    """True if `ref` is absolute or relative, as an import ref must be.

    A bare ref would need resolving by the very algorithm the import feeds, and
    that circularity has no good answer.
    """
    return ref.startswith("/") or ref.startswith("./") or ref.startswith("../")


def parse_statement(raw: str) -> ImportStmt:
    """Parse one statement, or raise `ImportError_` (HMD014)."""
    if not isinstance(raw, str):
        raise ImportError_(f"import entry must be a string, got {type(raw).__name__}")

    m = _STMT_RE.match(raw.strip())
    if not m:
        raise ImportError_(f"expected 'from <ref> import <names>', got {raw.strip()!r}")

    ref = m.group("ref")
    if not is_qualified(ref):
        raise ImportError_(
            f"import ref {ref!r} must be absolute ('/x') or relative ('./x', '../x'), never bare"
        )

    names = m.group("names").strip()
    if names == "*":
        return ImportStmt(ref=ref, wildcard=True, bindings=(), raw=raw.strip())

    bindings: list[tuple[str, str]] = []
    for part in names.split(","):
        part = part.strip()
        if not part:
            raise ImportError_(f"empty name in import list: {raw.strip()!r}")
        if part == "*":
            raise ImportError_("'*' may not be combined with named imports")
        nm = _NAME_RE.match(part)
        if not nm:
            raise ImportError_(f"expected '<name>' or '<name> as <alias>', got {part!r}")
        name = nm.group("name")
        bindings.append((name, nm.group("alias") or name))

    return ImportStmt(ref=ref, wildcard=False, bindings=tuple(bindings), raw=raw.strip())

File: tools/hmd-ts-core/src/imports.ts
/**
 * The `from <ref> import <names>` mini-grammar (HMD-0001 §5.3) — a port of
 * `tools/hmd/src/hypermarkdown/imports.py`.
 */

import type { ImportStmt } from "./model.js";

const STMT_RE = /^from\s+(\S+)\s+import\s+([\s\S]+?)\s*$/;
const NAME_RE = /^([^\s/]+)(?:\s+as\s+([^\s/]+))?$/;

/** A malformed import statement (HMD014). */
export class ImportSyntaxError extends Error {}

/**
 * True if `ref` is absolute or relative, as an import ref must be.
 *
 * A bare ref would need resolving by the very algorithm the import feeds, and
 * that circularity has no good answer.
 */
export function isQualified(ref: string): boolean {
  return ref.startsWith("/") || ref.startsWith("./") || ref.startsWith("../");
}

export function parseStatement(raw: unknown): ImportStmt {
  if (typeof raw !== "string") {
    throw new ImportSyntaxError(`import entry must be a string, got ${typeName(raw)}`);
  }

  const trimmed = raw.trim();
  const m = STMT_RE.exec(trimmed);
  if (!m) {
    throw new ImportSyntaxError(
      `expected 'from <ref> import <names>', got ${JSON.stringify(trimmed)}`,
    );
  }

  const ref = m[1]!;
  if (!isQualified(ref)) {
    throw new ImportSyntaxError(
      `import ref ${JSON.stringify(ref)} must be absolute ('/x') or relative ('./x', '../x'), never bare`,
    );
  }

  const names = m[2]!.trim();
  if (names === "*") {
    return { ref, wildcard: true, bindings: [], raw: trimmed };
  }

  const bindings: Array<readonly [string, string]> = [];
  for (const chunk of names.split(",")) {
    const part = chunk.trim();
    if (!part) {
      throw new ImportSyntaxError(`empty name in import list: ${JSON.stringify(trimmed)}`);
    }
    if (part === "*") {
      throw new ImportSyntaxError("'*' may not be combined with named imports");
    }
    const nm = NAME_RE.exec(part);
    if (!nm) {
      throw new ImportSyntaxError(
        `expected '<name>' or '<name> as <alias>', got ${JSON.stringify(part)}`,
      );
    }
    bindings.push([nm[1]!, nm[2] ?? nm[1]!] as const);
  }

  return { ref, wildcard: false, bindings, raw: trimmed };
}

function typeName(value: unknown): string {
  if (value === null) return "NoneType";
  if (Array.isArray(value)) return "list";
  return typeof value;
}
